man/man1/ccdppm.1
.TH CCDPPM 1NEMO "19 July 2020"
.SH NAME
ccdppm \- convert image to a PPM file, with optional color cable and ASINH scaling
.SH SYNOPSIS
\fBccdppm\fP [parameter=value]
.SH DESCRIPTION
\fBccdppm\fP converts a NEMO \fIimage(5NEMO)\fP to an 8/24bit PPM file.
A color table can be optionally added.
.PP
ppm files are probably the most versatile output graphics 
format, since it is very easy to write (no special library
is needed) and they can be easily converted to other formats
(gif, jpg,....) and are understood by most image processing
systems. See e.g. the \fIImageMagick(1)\fP toolset.
.PP
Although the \fPpower=\fP keyword can provide a simple
gamma factor type scaling, for astronomical data with a lot of pixels
with noise and some weak signal (positive, or negative!)
it is often useful
to use the "asinh" scaling, so clearly see both the structure
near the noise, as well as all the signal. Use the \fBmean=\fP
and \fPsigma=\fP parameters to activate the ASINH scaling.
.nf
     Out = slope*asinh((In-mean)/sigma))
.fi
where the \fIslope\fP is computed such that the minimum and maximum
are 0 and 1 resp.  Use \fIccdstat(1NEMO)\fP to find good values for
the mean and sigma, with its optional keyword \fBrobust=t\fP.
.PP
For certain types of objects (e.g. galaxies) the
ASINH scaling can be enhanced by also using power=0.5.

.SH "PARAMETERS"
.so man1/parameters
.TP 20
\fBin=\fP
Input image filename. Must be a \fIimage(5NEMO)\fP. Currently the
3rd dimension is ignored, and for datacubes the first slice is
taken. See \fIccdsub(5NEMO)\fP to select a plane. No default.
.TP
\fBout=\fP
Output PPM filename. No default.
.TP
\fBmin=\fP
Override data minimum used in scaling. By default it will look for
the minimum.
.TP
\fBmax=\fP
Override data maximum used in scaling. By default it will look for
the minimum.
.TP
\fBbad=\fP
Use this as masking value to ignore data. By default no data is 
considered bad.
.TP
\fBpower=\fP
The gamma factor \fBpower\fP applied to the input \fBx\fP,
i.e. \fIy = ((x-xmin)/(xmax-xmin))**power\fP. 
Use \fIccdmath(1NEMO)\fP for other transfer functions..  [1].
.TP
\fBmean=\fP
If supplied (and power= as well),
this will activate the ASINH scaling.
[not supplied]
.TP
\fBsigma=\fP
If supplied (and mean= as well),
this will activate the ASINH scaling. [not supplied]
.TP
\fBlut=\fP
optional selection of a lut from NEMODAT/lut. Currently the full path
to the file must be given. By default it will be using a grey scale
(red=green=blue).
.TP
\fB8bit=t|f\fP
Force up to 256 discrete colors. This can be useful if you want to convert
the normally 24bit ppms to 8bit images and not 'loose' any colors
in the conversion.    [Default: false].

.SH "EXAMPLES"
Here is an example to take the 5th plane from a fits file into a PPM file:
.nf

   % fitsccd cube.fits - | ccdsub - - z=4 | ccdppm - plane4.ppm lut=$NEMODAT/lut/ronekers.lut

.fi
Converting to other image formats, while also resizing it (including upscaling):
.nf

   % convert p1.ppm -resize 256x256 p1.jpg
   % fitsccd fpC-000094-g2-0446.fit - | ccdppm - - power=0.2 | convert - try.jpg

.fi
And using \fIccdstat(1NEMO)\fP and some trial and error picking the right mean/sigma/power :
.nf

   % fitsccd fpC-000094-g2-0446.fit - | ccdppm - try.ppm mean=1073 sigma=150 power=0.4

.nf
.SH CAVEATS
Each pixel from the input image is converted to an output PPM. The \fIImageMagick(1)\fP toolset
can be used to upscale the image count, or convert to other formats,such as JPG, PNG etc.
.SH SEE ALSO
ccdrgb(1NEMO), fitstiff(1NEMO), lut(5NEMO)
.PP
ASINH scaling Lupton et al. (2004) - https://ui.adsabs.harvard.edu/abs/2004PASP..116..133L/abstract
.SH FILES
$NEMODAT/lut   :  repository of \fIlut(5NEMO)\fP color lookup tables.
.SH AUTHOR
Peter Teuben
.SH UPDATE HISTORY
.nf
.ta +1.0i +4.0i
17-Dec-03	V1.0 Created (100 year Wright first flight anniversary)	PJT
.fi


src/image/misc/ccdppm.c
/* 
 * CCDPPM:   create a PPM from a CCD image
 *
 *	17-dec-03  V1.0 created   (Wright's 100 year first flight anniversary!)
 *                  1.1 added 8bit=
 *
 *  TODO
 *    - handle LUT's that are not 256 in size (e.g. ds9 uses 200 mostly)
 *    - handle lut='sls' to '$NEMODAT/lut/sls.lut' type translation
 */

#include <stdinc.h>
#include <getparam.h>
#include <filestruct.h>
#include <image.h>
#include <table.h>

string defv[] = {
    "in=???\n       Input image filename",
    "out=???\n      Output PPM filename",
    "min=\n         Minimum override",
    "max=\n         Maximum override",
    "bad=\n         Use this as masking value to ignore data",
    "power=1\n      Gamma factor applied to input data",
    "mean=\n        Mean value for ASINH scaling",
    "sigma=\n       Sigma value for ASINH scaling",
    "lut=\n         optional selection of a lut from NEMODAT/lut",
    "8bit=t\n       8bit?  (or 24bit) [not implemented]",
    "VERSION=1.3\n  19-jul-2020 PJT",
    NULL,
};

string usage="convert image to a PPM file, with optional color cable";

string	infile, outfile;       		/* file names */
stream  instr, outstr;			/* file streams */
imageptr iptr=NULL;			/* will be allocated dynamically */
int    nx,ny,nz;			/* actual size of map */

#define MAXCOL   256

int  ncol;
real  Red[MAXCOL],  Green[MAXCOL],  Blue[MAXCOL];
real dRed[MAXCOL], dGreen[MAXCOL], dBlue[MAXCOL];

void get_rgb(real data, real min, real max, int *R, int *G, int *B, bool);
int  get_lut(string, int, real*, real*, real*);

void nemo_main()
{
  int  i, j, red, green, blue;
  real x, xmin, xmax, bad, low, power, mean, sigma, slope;
  bool Qmin, Qmax, Qbad, Q8bit, Qasinh;
  
  instr = stropen (getparam("in"), "r");
  read_image (instr,&iptr);
  strclose(instr);
  outstr = stropen(getparam("out"), "w");
  Q8bit = getbparam("8bit");
  power = getrparam("power");
  Qasinh = hasvalue("mean") && hasvalue("sigma");
  if (Qasinh) {
    mean = getrparam("mean");
    sigma = getrparam("sigma");
  }

  if (!Q8bit) warning("24bit not working yet");
  
  nx = Nx(iptr);	
  ny = Ny(iptr);
  Qmin = hasvalue("min");
  if (Qmin) xmin = getdparam("min");
  Qmax = hasvalue("max");
  if (Qmax) xmax = getdparam("max");
  Qbad = hasvalue("bad");
  if (Qbad) bad = getdparam("bad");

  if (hasvalue("lut"))
    ncol = get_lut(getparam("lut"),MAXCOL,Red,Green,Blue);
  else {
    ncol = 256;
    for (i=0; i<ncol; i++) {
      Red[i] = Green[i] = Blue[i] = i/255.0;
      dRed[i] = dGreen[i] = dBlue[i] = 1.0/255.0;
    }
  }

  if (!Qmin || !Qmax) {                      /* loop over data to get xmin, xmax */
    if (!Qmin) xmin = MapValue(iptr,0,0);
    if (!Qmax) xmax = MapValue(iptr,0,0);
    for (j=0; j<ny; j++) {
      for (i=0; i<nx; i++) {
	x = MapValue(iptr,i,j);
	if (!Qmin && x<xmin) xmin=x;
	if (!Qmax && x>xmax) xmax=x;
      }
    }
    dprintf(0,"Auto-scaling, xmin=%g xmax=%g\n",xmin,xmax);
  }
  
  if (Qbad) {                              /* loop over data to flag bad data */
    dprintf(0,"Flag bad data\n");
    if (xmin<0)
      low = 2*xmin;
    else if (xmin>0)
      low = 0.5*xmin;
    else
      low = -xmax;
    for (j=0; j<ny; j++) {
      for (i=0; i<nx; i++) {
	x =  MapValue(iptr,i,j);
	if (Qbad && x==bad) MapValue(iptr,i,j) = low;
      }
    }
  }

  if (Qasinh) {                          /* if requested, apply the ASINH scaling */
    dprintf(0,"Using ASINH scaling with mean=%g sigma=%g\n",mean,sigma);
    slope = 1.0/asinh(xmax-xmin);
    for (j=0; j<ny; j++) {
      for (i=0; i<nx; i++) {
	x =  MapValue(iptr,i,j);
	MapValue(iptr,i,j) = asinh((x-mean)/sigma);
      }
    }
    xmin = 0;
    xmax = 1;
  }

  if (power != 1.0) {                   /* if requested, apply a gamma factor */
    dprintf(0,"Using gamma factor power=%g scaling\n",power);
    slope = 1.0/(xmax-xmin);    
    for (j=0; j<ny; j++) {
      for (i=0; i<nx; i++) {
	x =  MapValue(iptr,i,j);
	MapValue(iptr,i,j) = pow(slope*(x-xmin), power);
      }
    }
    xmin = 0;
    xmax = 1;
  }
  
  fprintf(outstr,"P6\n%d %d\n255\n",nx,ny);
  for (j=ny-1; j>=0; j--) {
    for (i=0; i<nx; i++) {
      x = MapValue(iptr,i,j);
      get_rgb(x,xmin,xmax,&red,&green,&blue,Q8bit);
      fputc(red,outstr);
      fputc(green,outstr);
      fputc(blue,outstr);
    }
  }
  strclose(outstr);
}

int  get_lut(string fn, int n, real *r, real *g, real *b)
{
  real *coldat[3];
  int i, ncol, colnum[3];
  stream instr = stropen(fn,"r");

  colnum[0] = 1;    coldat[0] = r;
  colnum[1] = 2;    coldat[1] = g;
  colnum[2] = 3;    coldat[2] = b;
  ncol = get_atable(instr,3,colnum,coldat,MAXCOL);
  if (ncol < 0) 
    error("Problem %d reading %s; MAXCOL=%d",ncol,fn,MAXCOL);
  for (i=1; i<ncol; i++) {
    dRed[i-1]   = Red[i]   - Red[i-1];
    dGreen[i-1] = Green[i] - Green[i-1];
    dBlue[i-1]  = Blue[i]  - Blue[i-1];
  }
  return ncol;
}


void get_rgb(real data, double min, double max, int *R, int *G, int *B, bool Q8bit)
{
  int i;
  real x;
  
  if (min==max) {
    *R = *G = *B = (data > max ? 255 : 0);    
    return;
  }

  if (!Q8bit) {
    x = (data-min)*255.0/(max-min);
    if (x<=0.0 || x>=255.0) {         /* out of bounds */
      i=  (x <= 0.0 ? 0 : 255);
      *R = (int) (Red[i] * 255.0);
      *G = (int) (Green[i] * 255.0);
      *B = (int) (Blue[i] * 255.0);
      return;
    }
    i = (int) x;
    *R = (int)  ((Red[i]   + (x-i)*dRed[i])*255.0);
    *G = (int)  ((Green[i] + (x-i)*dGreen[i])*255.0);
    *B = (int)  ((Blue[i]  + (x-i)*dBlue[i])*255.0);
  } else {
    i = (int) ((data-min)*255.0/(max-min));
    x = i;
    if (i<0) i=0;
    if (i>255) i=255;
    *R = (int) (Red[i] * 255.0);
    *G = (int) (Green[i] * 255.0);
    *B = (int) (Blue[i] * 255.0);
  }
  dprintf(1,"%g [%g %g] -> (%g,%d) %d %d %d\n",data,min,max,x,i,*R, *G, *B);
  dprintf(2,"%g %g   %g %g    %g %g\n",Red[i],dRed[i],  Green[i],dGreen[i], Blue[i],dBlue[i]);
}

void stern_special(real data, real min, real max, int *pRed, int *pGreen, int *pBlue)
{
  real max_min = max - min;
  real a, b, c, d;
  real x, y;

  if(max_min > 0.0){
    x = (data - min)/(max_min);

    /* ========== Red ============ */
    a = 18.214286*x;
    b = 1.2857143 - 5.2040816*x;
    c = a < b ? a : b;
    y = c < 0.0 ? x : c;
    *pRed = (int)(256.0*y);
    *pRed = *pRed >   0 ? *pRed :   0;
    *pRed = *pRed < 255 ? *pRed : 255;

    /* ========== Green ========== */
    *pGreen = (int)(256.0*x);
    *pGreen = *pGreen >   0 ? *pGreen :   0;
    *pGreen = *pGreen < 255 ? *pGreen : 255;

    /* ========== Blue =========== */
    a = 1.9921875*x;
    b = 3.1333333 - 4.25*x;
    c = 3.8059701*x - 2.8059701;
    d = a < b ? a : b;
    y = d > c ? d : c;
    *pBlue = (int)(256.0*y);
    *pBlue = *pBlue >   0 ? *pBlue :   0;
    *pBlue = *pBlue < 255 ? *pBlue : 255;
  }
  else{
    *pRed = *pGreen = *pBlue = (data > max ? 255 : 0);
  }

  return;
}

